### DISCUSSION TITLE
add analog clock

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#47] [Ideas]
Is is possible to run external scripts in specific layout,

For example in my newly created screensaver layout, I would like to display an analog clock like this one from luma example:
https://github.com/rm-hull/luma.examples/blob/master/examples/clock.py

How could I implement that? (callback function that just call the script?)

**Comment by mattblovell** [comment]
> How could I implement that? (callback function that just call the script?)

Just calling the script won't by itself actually suffice.  How would the external script "know" how to render to the same ImageDraw object as the running kodi_panel script?  You would have to arrange for some mechanism to transfer the image(s) generated from that program/script *back* to the Pillow object that kodi_panel ultimately renders.

Once the flow of the program enters one of your callback functions (e.g., some ``element_*`` rendering function that you've defined), you have full control to do, well, anything that Python is capable of doing.  So, if desired, you could use any of the approaches discussed on this page:

**6 ways to call external command in python**
https://www.newbedev.com/python/howto/6-ways-to-call-external-command-in-python/

What exactly happens subsequently is the tricky part.  (Also note that getting another program up and running definitely takes time.)

As your example is a luma.example script, you could just leverage the code that is responsible for actually drawing the clock.  That's effectively most of ``main()``, starting from Line 35 onwards:

https://github.com/rm-hull/luma.examples/blob/0eab29e331eb77270143e4c93cfd745b19be7099/examples/clock.py#L35-L62

That code sits inside a ``while True:`` forever loop (which should sound familiar).  The rest of the loop just gets the current date and time.

The drawing functionality could certainly be wrapped inside a new callback function (defined in your startup script), getting the current time from a Kodi InfoLabel.

**Reply by mattblovell** [reply]
See kodi_panel v1.46's release notes for the analog clock.

https://github.com/mattblovell/kodi_panel/releases/tag/v1.46

The more-general suggestion of being able to run arbitrary external scripts and capture their graphical output is much harder to accomplish.

**Comment by nico1080** [comment]
I confirm, it work, I also edited the title to mention anly analog clock

In my fork, I have also tweaked the clock to make it bigger and add mark for each hours aa0e64914a3c332e5784971e967624c109d20d61

It will require a little more work to have great tunning, but I will do it after I received my lcd display, and post picture then.
